pkg/dns/libvirt/provider.go

```diff
@@ -0,0 +1,114 @@
+package libvirt
+
+import (
+	"encoding/xml"
+	"net"
+	"strings"
+
+	libvirt "github.com/libvirt/libvirt-go"
+
+	configv1 "github.com/openshift/api/config/v1"
+	iov1 "github.com/openshift/cluster-ingress-operator/pkg/api/v1"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+)
+
+var (
+	// _ dns.Provider = &Provider{}
+	log = logf.Logger.WithName("entrypoint")
+)
+
+type Provider struct {
+	// config is required input.
+	config Config
+}
+
+type Config struct {
+	Cluster string
+	Domain  string
+	Url     string
+}
+type Host struct {
+	XMLName   xml.Name `xml:"host"`
+	Ip        string   `xml:"ip,attr"`
+	DnsRecord string   `xml:"hostname"`
+}
+
+type action = libvirt.NetworkUpdateCommand
+
+const (
+	upsertAction action = 3
+	deleteAction action = 2
+)
+
+// New creates (but does not start) a new operator from configuration.
+func New(config Config) (*Provider, error) {
+	provider := &Provider{
+		config: config,
+	}
+	return provider, nil
+}
+
+func (p *Provider) Ensure(record *iov1.DNSRecord, zone configv1.DNSZone) error {
+	return p.change(record, zone, upsertAction)
+}
+
+func (p *Provider) Delete(record *iov1.DNSRecord, zone configv1.DNSZone) error {
+	return p.change(record, zone, deleteAction)
+}
+
+// Change methods will perform an action on a record.
+func (p *Provider) change(record *iov1.DNSRecord, zone configv1.DNSZone, action action) error {
+	// Create a new connections to qemu
+	conn, err := libvirt.NewConnect(p.config.Url)
+	if err != nil {
+		log.Error(err, "failed to connect qemu")
+	}
+
+	// List all networks
+	networks, err := conn.ListAllNetworks(2)
+	if err != nil {
+		log.Error(err, "failed to get networks")
+	}
+
+	// Lookup app domain's ip
+	ips, err := net.LookupIP(p.config.Domain)
+	if err != nil || len(ips) == 0 {
+		log.Error(err, "failed to lookup app domain's IPs")
+	}
+	log.Info("Domain nslookup", "IPs", ips)
+
+	// Find the correct network and update a DNS record
+	for _, network := range networks {
```

Comment:
you don't need to manually find the network. You can use the `LookupNetworkByName` function:  https://github.com/libvirt/libvirt-go/blob/master/connect.go#L1159